### PR TITLE
Prohibit TLS 1.3 middlebox compatibility mode

### LINKED ISSUE #3594: Prohibit TLS middlebox compatibility mode
This came up in interop testing.  This mode causes the TLS stack to generate a ChangeCipherSpec message, which caused our server to crash.  That's a bug on our end, but it's still a problem.

We need to decide whether this mode is acceptable.  Right now, we only say that it is "redundant".  A prohibition this mode would be easier all around.  It would avoid spending 32 bytes for both endpoints too.

Unfortunately, this is a design change.

### PR DESCRIPTION
**Pull request by martinthomson**
Closes #3594.

### CODE REVIEW

File: draft-ietf-quic-tls.md [hunk 1]
```diff
@@ -1631,6 +1632,21 @@ PROTOCOL_VIOLATION.
...
+a workaround for bugs in some middleboxes. The TLS 1.3 middlebox compatibility
+mode involves setting the legacy_session_id field to a 32-byte value in the
+ClientHello and ServerHello, then sending a change_cipher_spec record. Both
+field and record carry no semantic content and are ignored.
+
+This mode has no use in QUIC as it only applies to middleboxes that interfere
+with TLS over TCP. A client MUST NOT request the use of the TLS 1.3
+compatibility mode. A server MUST treat the receipt of a TLS ClientHello that
+requests the use of the TLS 1.3 middlebox compatibility mode as a connection
+error of type PROTOCOL_VIOLATION.
```
Comment (kazuho) [inline draft-ietf-quic-tls.md:1647]: Use of MUST NOT for client is fine, but I would prefer using SHOULD or MAY on the server-side.

In a server-side TLS 1.3 stack, it is entirely reasonable to implement Compatibility Mode at the record layer in a reactive manner. In your record layer, when you receive the handshake traffic secret from the handshake layer, you inject CCS depending on the length of SH.legacy_session_id that has been retained.

Use of MUST NOT for the server breaks TLS stacks implemented that way.

Comment (marten-seemann) [reply draft-ietf-quic-tls.md:1647]: > In your record layer, when you receive the handshake traffic secret from the handshake layer, you inject CCS depending on the length of SH.legacy_session_id that has been retained. At the moment, there's nothing that has to be done at the handshake layer on the server-side.

Assume that you're not validating the empty `legacy_session_id` on the server side, how would you prevent the injection of CCS then?

Comment (martinthomson) [reply draft-ietf-quic-tls.md:1647]: I can do SHOULD.

The way our stack works is that we let TLS consume the ClientHello and then ask it for a response.  That produces a CCS if the client asked for one.  We turn that into an error (we used to crash).  Any implementation could, as I think Kazuho suggests, just drop the CCS, which might be easier in cases where your interface only passes handshake records.  We have a generic record callback, so detecting record type 20 is easy.

**Comment by marten-seemann** [comment]
Should we use a more specific error code than PROTOCOL_VIOLATION? Something in the 0x1XX range?

**Comment by martinthomson** [comment]
I couldn't see something more applicable.  You could fall back to a generic TLS alert, like handshake_failure, but the error is really a QUIC one.

**Comment by kazuho** [comment]
@marten-seemann 
> Assume that you're not validating the empty legacy_session_id on the server side, how would you prevent the injection of CCS then?

As @martinthomson says, our code that emits ChangeCipherSpec would check if the underlying protocol is the TLS record protocol. If that's something else, a TLS alert is generated. See https://github.com/h2o/picotls/blob/master/lib/picotls.c#L1480-L1484.

@martinthomson 
> You could fall back to a generic TLS alert, like handshake_failure, but the error is really a QUIC one.

I'm not sure if I share that view. We already have certain requirements on how TLS is used, and we use TLS error codes to communicate errors (e.g., use of unexpected_message when a TLS Key Update message is received).

FWIW, the code above in picotls sends unexpected_message (I can change that for sure, I just merged it).

**Comment by martinthomson** [comment]
> the code above in picotls sends unexpected_message 

That seems fine, but in my view the fault was with the ClientHello, not the CCS.

**Comment by kazuho** [comment]
> That seems fine, but in my view the fault was with the ClientHello, not the CCS.

I agree that the fault is with ClientHello, it's merely a coincidence that we detect that problem when trying to build particular portion of the response. In case of picotls, return of this error will be propagated to the application alongside a partially built buffer, so QUIC stacks using picotls would just send the TLS error code.

File: draft-ietf-quic-tls.md [hunk 2]
```diff
@@ -213,6 +209,11 @@ own key update mechanism {{key-update}}.
 ~~~
 {: #tls-full title="TLS Handshake with 0-RTT"}
 
+{{tls-full}} omits the EndOfEarlyData message, which is not used in QUIC; see
+{{remove-eoed}}. Likewise, neither ChangeCipherSpec nor KeyUpdate messages are
+used by QUIC. ChangeCipherSpec is redundant in TLS 1.3; see {{compat-mode}}.
+QUIC has its own key update mechanism; see {{key-update}}.
```
Comment (ekr) [inline draft-ietf-quic-tls.md:215]: Worth noting that it's not possible to send CCS anywya?

File: draft-ietf-quic-tls.md [hunk 3]
```diff
@@ -1631,6 +1632,21 @@ PROTOCOL_VIOLATION.
...
+Appendix D.4 of {{!TLS13}} describes an alteration to the TLS 1.3 handshake as
+a workaround for bugs in some middleboxes. The TLS 1.3 middlebox compatibility
+mode involves setting the legacy_session_id field to a 32-byte value in the
+ClientHello and ServerHello, then sending a change_cipher_spec record. Both
+field and record carry no semantic content and are ignored.
+
+This mode has no use in QUIC as it only applies to middleboxes that interfere
+with TLS over TCP. A client MUST NOT request the use of the TLS 1.3
+compatibility mode. A server SHOULD treat the receipt of a TLS ClientHello that
+requests the use of the TLS 1.3 middlebox compatibility mode as a connection
```
Comment (ekr) [inline draft-ietf-quic-tls.md:1646]:
```suggestion
with a non-empty legacy_session_id field as a connection
```

Comment (ekr) [reply draft-ietf-quic-tls.md:1646]: This way you don't need to dig into D.4 to see what the signal is.

File: draft-ietf-quic-tls.md (hunk 2 above)
Comment (martinthomson) [reply draft-ietf-quic-tls.md:215]: Not here, but I'll do that below.

**Comment by huitema** [comment]
I observe that implementation of this PR is causing interop failures during tests. Some implementations have been updated, some have not been. I recently updated Picoquic to use the latest version of Picotls, and the results are mixed: Handshake with Quant and Quickly that were failing before the update are now succeeding; handshakes with Quiche that were succeeding are now failing. This is clearly a case of spec confusion, with some stacks rushing to test this PR and some sticking to the stated spec. How about clearing that confusion, updating the spec and releasing draft-28?

**Comment by ianswett** [comment]
@huitema that is the plan as soon as the chairs declare consensus.

**Comment by LPardue** [comment]
Yep. The consensus call for this ends tomorrow and there seems to be no push back so far...

**Pull request merged**